functions/logDonation.js

```diff
@@ -15,164 +15,85 @@ const gmaps = new Client({});
 const DEFAULTCPL = 1.0;
 exports.logDonation = functions.https.onRequest(async (req, res) =>{
   if (!req.body) {
-    res.status(501).send('no data supplied!');
-    return;
+    return res.status(400).send('no data supplied!');
   }
+  try {
+    const event = req.body;
 
-  const splitString = req.body.campaignID.split('|');
-  if (splitString.length < 3) {
-    return res.status(400).send(`Missing required fields in campaignID: <country>|<campaign>|<referralSource>`);
-  }
-
-  let country = splitString[0] || 'MISSING'
-  let campaign = splitString[1] || 'MISSING'
-  let referralSource = splitString[2] || 'MISSING'
+    const splitString = (event.campaignID||'').split('|');
 
-  let amount = Number(req.body.amount);
-  if (req.body.coveredByDonor) {
-    amount = amount - Number(req.body.coveredByDonor);
-  }
-  const params = {
-    firstName: req.body.firstName,
-    email: req.body.email,
-    timestamp: admin.firestore.Firestore.Timestamp.now(),
-    amount: amount,
-    frequency: req.body.frequency,
-    campaignID: campaign,
-    country: country,
-    referralSource: referralSource,
-  };
-  for (param in params) {
-    if (params[param] && params[param] === 'MISSING') {
-      params['needsAttention'] = true;
-    } else if (!params[param]) {
-      params[param] = 'MISSING';
-      params['needsAttention'] = true;
+    let campaign = splitString[0] || 'MISSING'
+    let country = splitString[1] || 'MISSING'
+    let referralSource = splitString[2] || 'MISSING'
+    let amount = Number(event.amount);
+    if (event.coveredByDonor) {
+      amount = amount - Number(event.coveredByDonor);
     }
+    const params = {
+      firstName: event.firstName,
+      email: event.email,
+      amount: amount,
+      frequency: event.frequency,
+      campaignID: campaign,
+      country: country,
+      referralSource: referralSource,
+    };
+    for (param in params) {
+      if (params[param] && params[param] === 'MISSING') {
+        params['needsAttention'] = true;
+      } else if (!params[param]) {
+        params[param] = 'MISSING';
+        params['needsAttention'] = true;
+      }
+    }
+    const uid = await helpers.getOrCreateDonor(params.email);
+    params.sourceDonor = uid;
+    this.writeDonation(params)
+    const msg = {msg:'successfully handled payment', uid: uid};
+    return res.status(200).send({msg: msg, data: event});
+  } catch (err) {
+    const msg = {err: err};
+    console.error(`encountered an error handling payment: ${err}`);
+    return res.status(501).send({msg: msg, event: req.body})
```

Comment:
This should be a 500 error since it's most likely something internal to our system